docs: update troubleshooting pages (#242)
wasmCloud/wasmCloud.com: docs/developer/debugging/actors.mdx

---
title: "Actor Troubleshooting"
date: 2022-01-19T11:02:05+06:00
sidebar_position: 3
draft: false
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

:::info

> Before you follow this section, consider [configuring the log level](/docs/developer/debugging/host#changing-the-log-level) to `debug` in order to see the most information when debugging.

Reviewer: ```suggestion
Before you follow this section, consider [configuring the log level](/docs/developer/debugging/host#changing-the-log-level) to `debug` for more information while debugging.
```

:::

When developing actors, there are a few common errors that can cause your actor to fail to run or properly operate. Below we'll walk through a few of these scenarios with accompanying error messages.

### Failing to start an actor from a local registry over HTTP

The following error can happen when attempting to start an actor from a local registry server over HTTP: 

```console
2023-12-19T22:31:26.154791Z ERROR wasmcloud_host::wasmbus: failed to scale actor actor_ref=127.0.0.1:5000/v2/hello:0.1.0 err=failed to fetch actor

Caused by:
    0: failed to fetch actor under OCI reference `127.0.0.1:5000/v2/hello:0.1.0`
    1: failed to fetch OCI path
    2: failed to fetch OCI bytes
    3: error sending request for url (https://127.0.0.1:5000/v2/): error trying to connect: received corrupt message of type InvalidContentType
```

To fix this error, you need to set the `--allowed-insecure` [argument on the host](/docs/reference/host-config).

:::warning

> When accessing a local registry from within a docker compose network, you should use `registry:5000` (or the name of the registry container, if it differs) instead of `localhost:5000`. Even though port 5000 may be available on your host's network, networking works differently within docker, and you should use the name of the container instead of localhost. We do the same thing with `nats`, as you can see below.

Reviewer: ```suggestion
When accessing a local registry from within a docker compose network, you should use `registry:5000` (or the name of the registry container, if it differs) instead of `localhost:5000`. 

While port 5000 may be available on your host's network, default docker compose networking settings often require using the name of the container instead of `localhost`. We follow this practice when referring to `nats` below.
```

:::

### Failing to invoke an actor

#### Missing Capability Claims

> When a wasmCloud actor needs access to a capability provider, we require cryptographically signed claims embedded in that actor. The host runtime operates on a deny-by-default basis and require claims to be an allow-list to ensure our actors are secure. Adding these capability claims is mostly managed for you with `wash` when you build your actor and the list of capability claims in `wasmcloud.toml` included with any actor project.

Reviewer: ```suggestion
When a wasmCloud actor needs access to a capability provider, we require cryptographically signed claims embedded in that actor. The host runtime operates on a deny-by-default basis and require claims to be an allow-list to ensure our actors are secure. 

Adding capability claims is mostly managed for you with `wash` by listing the relevant capability claims in `wasmcloud.toml` when building your actor project.
```

> Consider the `KVCounter` example actor, which uses the `wasmcloud:httpserver` capability to receive HTTP requests, and the `wasmcloud:keyvalue` capability to issue requests to a keyvalue database. Omitting either of these capabilities from the actor's claims will cause errors at runtime, as the wasmCloud host will deny the invocation:

Reviewer: ```suggestion
Consider the `KVCounter` example actor, which uses:
- `wasmcloud:httpserver` capability to receive HTTP requests
- `wasmcloud:keyvalue` capability to issue requests to a keyvalue database

Omitting either of these capabilities from the `KVCounter` actor's `claims` list (in `wasmcloud.toml`) will cause errors at runtime, as the wasmCloud host will deny the invocation:
```

Caused by:
    actor does not have capability claim `wasmcloud:httpserver`
```

To fix this error, make sure the actor's `wasmcloud.toml` file includes the appropriate capability claims:

```toml
[actor]
claims = ["wasmcloud:httpserver"]
```

#### Missing Runtime Links

Invocations are only allowed between actors and providers that have been [linked](/docs/concepts/runtime-linking). Omitting a link will result in a runtime error when the host denies the invocation:

```
{"error":"Host send error failed to call `wasmcloud:bus/host.call`: failed to call linked provider: unknown actor call alias"}
```

To fix this error, make sure actors and providers are linked to each other appropriately. To inspect the current set of runtime links, run:

```bash
wash get links
```

> The command line wasmCloud shell (aka [wash](https://github.com/wasmCloud/wasmCloud/tree/main/crates/wash-cli)) attempts to safeguard against misspellings, though it is unable to determine if a particular link is between the intended actor and provider. If any of the actor ID, provider ID, link name, or contract ID are misspelled or omitted, then the host will deny the invocation.

Reviewer: ```suggestion
The command line wasmCloud shell (aka [wash](https://github.com/wasmCloud/wasmCloud/tree/main/crates/wash-cli)) attempts to safeguard against misspellings, though it is unable to determine if a particular link is between the intended actor and provider. 

If any of the actor ID, provider ID, link name, or contract ID are misspelled or omitted, then the host will deny the invocation.
```
